internal/encoder/vm_escaped/vm.go

```diff
@@ -3646,8 +3646,13 @@ func Run(ctx *encoder.RuntimeContext, b []byte, codeSet *encoder.OpcodeSet, opt
 				code = code.NextField
 				break
 			}
+			iface := ptrToInterface(code, p)
+			if code.Nilcheck && encoder.IsNilForMarshaler(iface) {
```

Comment:
Probably this condition will not be true in the test cases you added .
( you can see VM opcodes by `fmt.Println(code.CodeSet.Dump())` )